### ISSUE TITLE
System.IO.IOException: Sharing violation on path /storage/emulated/0/Android/data/com.rolsped.dev.TruckerApp/files/Pictures/temp/IMAG0053_1.jpg

### ISSUE THREAD

**Issue by ericbrunner** [jamesmontemagno/MediaPlugin#196]
Please fill out either the bug or feature request section and remove whatever section you are not using.

## Bug 

Version Number of Plugin: 2.6.3.6-beta
Device Tested On: HTC Desire 510
OS: Android 4.4 API-LEVEL 19
Simulator Tested On: No

### Expected Behavior
- Pick a Foto
- Gallery is displayed
- Select a photo
- Dispose MediaFile and delete temp file 
- Pick a Foto
- Gallery should be displayed

### Actual Behavior
- Pick a Foto
- Gallery is displayed
- Select a photo
- Dispose MediaFile and delete temp file 
- Pick a Foto
- Gallery Screen is black
- Exception thrown (AggregateException)

### Steps to reproduce the Behavior

- Pick a Foto
- Gallery is displayed
- Select a photo
- Dispose MediaFile and delete temp file 
- Pick a Foto
- Gallery Screen is black
- Exception thrown (AggregateException)

Exception:

Stacktrace on Gist: https://gist.github.com/ericbrunner/32566d18a9c89043825eb555ea2d7536

Code:

```
               SelectedMedia?.Dispose();

                if (SelectedMedia?.MediaState == MediaState.Uploaded)
                {
                    await DeleteFileAfterUploadAsync();
                }
```

```
        private async Task DeleteFileAfterUploadAsync()
        {
            //Platform specific file delete
            var truckerappMedia = SelectedMedia.FilePath;

            var platformFileHandler = DependencyService.Get<IFileHandling>();

            if (platformFileHandler != null)
                if (await platformFileHandler.FileExistsAsync(truckerappMedia))
                {
                    var deleted = await platformFileHandler.DeleteFileAsync(truckerappMedia);

                    if (!deleted)
                    {
                        await MetricsManagerHelper.Instance.SendErrorToApplicationInsightsAsync($"Photo couldn't be deleted: {truckerappMedia}");
                    }
                }
        }
```

**Comment by jamesmontemagno** [comment]
Did you follow: https://github.com/jamesmontemagno/MediaPlugin#android-n

Please don't post strack traces here, use Gist.github.com

**Comment by ericbrunner** [comment]
It worked prior to the time I added the `DeleteFileAfterUploadAsync ` to delete the picked photo. Otherwise when I don't delete that temp file I noticed that my data storage for that app increases and don't get purged. So a RAM memory leak occurs.

That is my setting

AndroidManifest.xml

```
<?xml version="1.0" encoding="utf-8"?>

<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.rolsped.dev.TruckerApp"
          android:installLocation="auto" android:versionCode="236" android:versionName="1.0">
  <uses-sdk android:minSdkVersion="16" />
  <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
  <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
  <uses-permission android:name="android.permission.BATTERY_STATS" />
  <uses-permission android:name="android.permission.BLUETOOTH" />
  <uses-permission android:name="android.permission.DEVICE_POWER" />
  <uses-permission android:name="android.permission.READ_PHONE_STATE" />
  <uses-permission android:name="android.permission.INTERNET" />
  <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
  <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
  <uses-permission android:name="android.permission.CAMERA" />
  <uses-permission android:name="android.permission.VIBRATE" />
  <uses-permission android:name="android.permission.FLASHLIGHT" />
  <application android:label="TruckerApp DEV" android:icon="@drawable/ic_launcher" android:largeHeap="true">
    <provider android:name="android.support.v4.content.FileProvider"
              android:authorities="com.rolsped.dev.TruckerApp.fileprovider" android:exported="false"
              android:grantUriPermissions="true">
      <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/file_paths_dev"></meta-data>
    </provider>
  </application>
</manifest>
```

file_path_dev.xml in Resources/xml

```
<?xml version="1.0" encoding="utf-8"?>
<paths xmlns:android="http://schemas.android.com/apk/res/android">
  <external-path name="my_images" path="Android/data/com.rolsped.dev.TruckerApp/files/Pictures" />
  <external-path name="my_movies" path="Android/data/com.rolsped.dev.TruckerApp/files/Movies" />
</paths>
```

**Comment by jamesmontemagno** [comment]
It could be that you are still trying to access things when deleting it. Have you tried to remove your deletion code and see if the library by itself works?

You may need to delete after the dispose of the image maybe....

**Comment by ericbrunner** [comment]
If I remove deletion code it works. I tried to put deletion after dispose but then it thus not work.

When I don't delete the picked photo , I will end up in growing app data storage ...

**Comment by jamesmontemagno** [comment]
My assumption is maybe it is something in your delete code that is causing this, not the plugin.

**Comment by ericbrunner** [comment]
I will try if I come in a situation where I delete the file prior to dispose the selected media and get back if I find something relevant ...

**Comment by ericbrunner** [comment]
I did some tests with my two different devices:

**Device**: HTC Desire 510 
**OS**: 4.4 , API-LEVEL 19

**Device**: HUAWEI P9
**OS**: 6.0, API-LEVEL 23

**1st Test: Take a photo**

I noticed that the taken photo appears in the public gallery in the HTC Desire 510 but on the HUAWEI P9 the photo is not stored in the public gallery. Note: both tests are with same code (see gist from previous post)!

**2nd Test: Pick that Photo (on the HTC, because on the HUAWEI P9 not in the public gallery)**

I could load that photo from the public gallery on the HTC. 

I sent that photo, then I disposed the mediafile and afterwards I deleted the temp file of the picked photo.

Then I executed `PickMediaAsync `again, Gallery appears,...

I select the same photo => Gallery Screen gets black and that exception from the initial reported issue is thrown...

Mmmmmm........... scratch my head ........ I guess in that case (on the HTC, API LEVEL 19) the Media Plugin "stores" that path to the "temp" picked foto file somehow and because I deleted that temp picked photo file I ran into the FileNotFound Exception.


In the meanwhile I decided to delete picked files "only" on Android devices that have at least API LEVEL 23 (which are most of my clients) like with that platform specific code. On iOS devices from Version 4s to 6 plus I (gladfully) had/have no issues:


   public async Task<bool> DeleteFileAsync(string filepath, MediaType mediaType)
        {
            bool deleted;

            try
            {
                if ((int) Build.VERSION.SdkInt >= 23 && (mediaType | MediaType.PickedPhoto) != 0)
                {
                    if (System.IO.File.Exists(filepath))
                    {
                        System.IO.File.Delete(filepath);
                        deleted = true;
                    }
                    else
                    {
                        deleted = false;
                    }
                }
                else
                {
                    deleted = true;
                }
            }
            catch (Exception ex)
            {
                await MetricsManagerHelper.Instance.SendExceptionToApplicationInsightsAsync(ex);
                deleted = false;
            }

            return deleted;
        } 
```
